Feature/1781 Twilio is integrated with app for event triggered notifications

harsohailB opened the pull request:
### What github issue is this PR for, if any?

Resolves #1781 

### What changed, and why?

* Added `Twilio Account SID`, `Twilio API Key SID`, `Twilio API Key Secret`, and `Twilio Phone Number` under the "Edit Organization" tab

These are given by Twilio in their console as follows:

<img width="400" alt="Screen Shot 2022-05-09 at 12 22 56 PM" src="https://user-images.githubusercontent.com/47438886/167472991-762df643-2830-472c-9ba7-725fb27ee8cd.png">

<img width="400" alt="Screen Shot 2022-05-09 at 5 19 10 PM" src="https://user-images.githubusercontent.com/47438886/167513830-684bc485-4a9e-4178-9263-20cf2a5f0b91.png">


A POST request like the following is sent to send an SMS which requires these three parameters:

* Changed `CasaOrg` model to store the three attributes

```
curl -X POST https://api.twilio.com/2010-04-01/Accounts/${TWILIO_ACCOUNT_SID}/Messages.json \
                --data-urlencode "Body=Hello from Twilio" \
                --data-urlencode "From=${TWILIO_PHONE_NUMBER}" \
                --data-urlencode "To=${RECIPIENT NUMBER}" \
                -u ${TWILIO_API_KEY_SID}:${TWILIO_API_KEY_SECRET}
```

### How will this affect user permissions?
- Volunteer permissions: n/a
- Supervisor permissions: n/a
- Admin permissions: n/a

### How is this tested? (please write tests!) 💖💪

* Wrote view tests for the "Edit Organization" page which now contains Twilio inputs
* Wrote requests tests to verify update requests for a `CasaOrg` (invalid phone number input not accepted)

### Screenshots please :)

![Screen Shot 2022-05-10 at 11 31 09 AM](https://user-images.githubusercontent.com/47438886/167688007-bed5f600-9d77-49ef-9f09-4afa1d33a0fe.png)

harsohailB committed:
create columns to save twilio auth data for casa orgs

harsohailB committed:
add casa org validator to validate phone number format and write tests for casa org update request

harsohailB committed:
remove casa org validator from testing

xihai01 commented on app/validators/casa_org_validator.rb:
```diff
@@ -0,0 +1,38 @@
+
+  private
+
+  def valid_twilio_phone_number(number, record)
```
@harsohailB Can this method be DRY'd up? It seems awfully similar to the phone number validation method for validators/user_validator.rb

harsohailB replied:
Will do!

harsohailB replied:
Abstracted out phone number validity logic into a helper named `PhoneNumberHelper`. Will create a ticket to point our previously worked on `UserValidator` and importers to this as well for code reuse.

xihai01 commented on db/migrate/20220508232038_add_columns_to_casa_orgs.rb:
```diff
@@ -0,0 +1,7 @@
+class AddColumnsToCasaOrgs < ActiveRecord::Migration[7.0]
+  def change
+    add_column :casa_orgs, :twilio_account_sid, :string
```
@harsohailB I'm wondering, but should these API credentials be hashed before storing them in the db? https://security.stackexchange.com/questions/180345/do-i-need-to-hash-or-encrypt-api-keys-before-storing-them-in-a-database

Maybe @compwron can give some suggestions.

compwron replied:
In general yes. Use https://guides.rubyonrails.org/active_record_encryption.html

harsohailB replied:
I will look into that and hash before storing!

harsohailB replied:
Encrypted the Twilio credentials. Added random key set for encryption in `config/credentials.yml.enc` locally, however, this is git ignored. This is why the tests are failing. @compwron How can we add such encryption key sets in other environments?

harsohailB replied:
Resolved by having a separate credential file for each environment.

xihai01 commented on spec/requests/casa_org_spec.rb:
```diff
@@ -2,7 +2,7 @@
 RSpec.describe "CasaOrg", type: :request do
   let(:casa_org) { build(:casa_org) }
-  let(:valid_attributes) { {name: "name", display_name: "display_name", address: "address"} }
+  let(:valid_attributes) { {name: "name", display_name: "display_name", address: "address", twilio_account_sid: "fyxpz5naqnir3ftopvxuzev6ir48xb4jmc", twilio_auth_token: "gukogx4k99885clw6j7ucd62hgzj5w4p", twilio_phone_number: "+12223334444"} }
```
@harsohailB just making sure, but these API credentials are fake right?

harsohailB replied:
Yes they are fake

xihai01 reviewed (approved):
awesome stuff! 😄 Left a few comments for you. I also have a suggestion. According to the Twilio docs, they recommend using API keys since they aren't attached to the main account and can be easily issued and revoked: https://www.twilio.com/docs/usage/rest-api-best-practices#control-access-to-your-accounts

compwron commented:
Yes, API keys- and stored as env variables

harsohailB committed:
change db fields to store api key information rather than auth tokens

harsohailB committed:
update tests according to api key changes

harsohailB committed:
add phone number helper to abstract out checking phone number validity

harsohailB commented:
> awesome stuff! 😄 Left a few comments for you. I also have a suggestion. According to the Twilio docs, they recommend using API keys since they aren't attached to the main account and can be easily issued and revoked: https://www.twilio.com/docs/usage/rest-api-best-practices#control-access-to-your-accounts

Refactored to use API keys instead. Updated the description (curl request + pics) on how I understand they work.

harsohailB committed:
encrypt twilio credentails

harsohailB committed:
fix linting issues

harsohailB committed:
add tests for phone number helper

7riumph reviewed (commented):
Edit: misread an above comment/am not seeing multiple credentials in the codebase

7riumph reviewed (approved):
Nice! @xihai01 has great comments too 🔥

xihai01 commented:
@harsohailB Don't forget to update the screenshot

xihai01 reviewed (approved):
LGTM

compwron commented:
Failing tests are 
```
rspec ./spec/system/casa_org/edit_spec.rb:96 # casa_org/edit can update show_driving_reimbursement flag
rspec ./spec/system/casa_org/edit_spec.rb:106 # casa_org/edit can upload a logo image
rspec ./spec/requests/casa_org_spec.rb:36 # CasaOrg as an admin PATCH /update with valid parameters doesn't revert logo to default if non logo details are updated
rspec ./spec/requests/casa_org_spec.rb:22 # CasaOrg as an admin PATCH /update with valid parameters updates the requested casa_org
rspec ./spec/requests/casa_org_spec.rb:44 # CasaOrg as an admin PATCH /update with valid parameters redirects to the casa_org
rspec ./spec/requests/casa_org_spec.rb:50 # CasaOrg as an admin PATCH /update with valid parameters also responds as json
```

harsohailB committed:
add encryption credentials file

harsohailB committed:
add rails master key secret location to github workflows

xihai01 reviewed (commented):
great!

harsohailB committed:
move master key env to all steps in github workflows

harsohailB committed:
create seperate credential files based on environment

harsohailB committed:
update production creds file that can be decrypted by master key

harsohailB commented:
During production deployment, we will need to put `RAILS_MASTER_KEY` as an environment variable in Heroku so the `production.yml.enc` file can be decrypted.

compwron reviewed (approved):
:)

compwron committed:
Merge branch 'main' into feature/1781

xihai01 commented:
@harsohailB tests are failing. Seems like you need to run a migration
![Screenshot from 2022-05-14 09-51-07](https://user-images.githubusercontent.com/86758440/168428667-3267e527-8d77-4ce4-9581-328080e11f6f.png)

harsohailB committed:
run the migrations required after main merge

harsohailB committed:
Merge branch 'main' into feature/1781
